src/lint/config/index.ts

```diff
@@ -1,14 +1,35 @@
+import { safeLoad, YAMLException } from "js-yaml";
 import ConfigSchema, { Config, RulesPreset } from "./schema";
 import presets, { defaultPreset, defaultPresetName } from "../presets";
 
 export const configFileName = ".gitlintrc";
 export const configFileExtensions = ["", ".yaml", ".yml"];
 
+export type FileLoader = (path: string) => Promise<string | null>;
+
+export async function discoverConfig(
+  fileLoader: FileLoader
+): Promise<Config | null> {
+  let config = null;
+  for (const extension of configFileExtensions) {
+    const path = configFileName + extension;
+    const configString = await fileLoader(path);
+    if (configString === null) continue;
+    try {
+      config = safeLoad(configString);
+    } catch (err) {
+      if (err === YAMLException) continue;
```

Comment:
I don't think we should use an equality check? `err instanceof YAMLException` seems correct.

Also, this function silently swallow all errors from configuration YAML parsing and validation. Callers can only distinguish between 'a valid config' and 'config is not found or invalid', which is not helpful to user.

I'll suggest wrapping the errors into a custom `Error` sub-class, so caller can catch an error with message like `cannot parse '.gitlintrc.yaml': invalid token at line 9`, or `invalid configuration '.gitlintrc': preset is not correct`.